src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,369 @@
+          while (self->server_unavailable_timeout_time_.has_value() && !self->shutdown_) {
+            self->CheckChannelStatus(false);
+            std::this_thread::sleep_for(std::chrono::milliseconds(
+                self->check_channel_status_interval_milliseconds_));
+          }
+          if (self->shutdown_) {
+            callback(Status::Disconnected("GRPC client has been shutdown."), Reply());
+            delete executor;
+          } else {
+            executor->Retry();
+          }
+        } else {
+          bool shutdown = false;
+          {
+            absl::MutexLock lock(&self->mu_);
+            if (!self->shutdown_) {
+              self->pending_requests_bytes_ += request_bytes;
+              auto timeout = timeout_ms == -1
+                                 ? absl::InfiniteFuture()
+                                 : absl::Now() + absl::Milliseconds(timeout_ms);
+              self->pending_requests_.emplace(timeout,
+                                              std::make_pair(executor, request_bytes));
+            } else {
+              shutdown = true;
+            }
+          }
+          if (shutdown) {
+            // Run the callback outside of the lock.
+            callback(Status::Disconnected("GRPC client has been shutdown."), Reply());
+            delete executor;
+          } else if (!self->server_unavailable_timeout_time_.has_value()) {
+            self->server_unavailable_timeout_time_ =
+                absl::Now() + absl::Seconds(self->server_unavailable_timeout_seconds_);
+            self->SetupCheckTimer();
+          }
+        }
+      }
+    };
+    auto operation = [prepare_async_function,
+                      &grpc_client,
+                      call_name,
+                      request,
+                      operation_callback,
+                      timeout_ms]() {
+      grpc_client.template CallMethod<Request, Reply>(
+          prepare_async_function, request, operation_callback, call_name, timeout_ms);
+    };
+    executor->Execute(std::move(operation));
+  }
+
+  void Shutdown() {
+    absl::MutexLock lock(&mu_);
+    if (!shutdown_.exchange(true)) {
+      // First call to shut down this GRPC client.
+      timer_->cancel();
+
+      // Fail the pending requests.
+      while (!pending_requests_.empty()) {
+        auto iter = pending_requests_.begin();
+        Executor *executor = iter->second.first;
+        // Make sure the callback is executed in the io context thread.
+        io_context_.post(
+            [executor]() {
+              executor->Abort(Status::Disconnected("GRPC client is shut down."));
+              delete executor;
+            },
+            "RetryableGrpcClient.Shutdown");
+        pending_requests_.erase(iter);
+      }
+      pending_requests_bytes_ = 0;
+    } else {
+      RAY_LOG(DEBUG) << "GRPC client has already been shut down.";
+    }
+  }
+
+  // Return the number of pending requests waiting for retry.
+  size_t NumPendingRequests() {
```

Comment:
`const` specifier?

Reply:
`absl::MutexLock lock(&mu_);` prevents it.

Reply:
Should use `mutable` for mutex.